gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+import json
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+from torchtyping import TensorType
+
+from gflownet.envs.base import GFlowNetEnv
+from gflownet.envs.crystals.composition import Composition
+from gflownet.envs.crystals.lattice_parameters import LatticeParameters
+from gflownet.envs.crystals.spacegroup import SpaceGroup
+from gflownet.utils.crystals.constants import TRICLINIC
+
+
+class Stage(Enum):
+    """
+    In addition to encoding current stage, values of this enum are used for padding individual
+    component environment's actions (to ensure they have the same length for tensorization).
+    """
+
+    COMPOSITION = -2
+    SPACE_GROUP = -3
+    LATTICE_PARAMETERS = -4
+
+
+class Crystal(GFlowNetEnv):
+    """
+    A combination of Composition, SpaceGroup and LatticeParameters into a single environment.
+    Works sequentially, by first filling in the Composition, then SpaceGroup, and finally
+    LatticeParameters.
+    """
+
+    def __init__(
+        self,
+        composition_kwargs: Optional[Dict] = None,
+        space_group_kwargs: Optional[Dict] = None,
+        lattice_parameters_kwargs: Optional[Dict] = None,
+        **kwargs,
+    ):
+        self.composition_kwargs = composition_kwargs or {}
+        self.space_group_kwargs = space_group_kwargs or {}
+        self.lattice_parameters_kwargs = lattice_parameters_kwargs or {}
+
+        self.composition = Composition(**self.composition_kwargs)
+        self.space_group = SpaceGroup(**self.space_group_kwargs)
+        # We initialize lattice parameters here with triclinic lattice system to access
+        # all the methods of that environment, but it will have to be reinitialized using
+        # proper lattice system from space group once that is determined.
+        # Triclinic was used because it doesn't force any initial starting angles.
```

Comment:
This is a bit annoying no? How about we make `LatticeParameters` be unconstrained by default? Would that solve the mentioned stateless issues with the lattice parameters?

Reply:
Stateless thing is a separate issue that wouldn't be solved by this. Yes, it is annoying, but I don't think making `LatticeParameters` unconstrained will help - the thing is that `LatticeParameters` just has to have constraints to work properly (to determine the source, masks etc), and if anything, removing the need for the constraint will make it work like using the triclinic system anyway, while making the `LatticeParameters` have additional edge case to consider internally.